### PR TITLE
Add a data setter

### LINKED ISSUE #779: Add a way of setting `.data`
### Describe the feature

I frequently have the case when working with DKIST data that I want to in-place compute the dask array, i.e. this:
```python
spatial_avg = visp.rebin((-1, 1, -1), ).squeeze()
spatial_avg._data = spatial_avg.data.compute()
```
it would be nice if I didn't have to use `._data`.

### Proposed solution

Add a setter for `.data` which validates the new array is the same shape as the old array.

### PR DESCRIPTION
**Pull request by Cadair**
fixes #779 

Main open question is if we should allow changing the unit by setting the data with a quantity.

### CODE REVIEW

File: ndcube/conftest.py [hunk 1]
```diff
@@ -575,7 +575,6 @@ def ndc(request):
 # NDCubeSequence Fixtures
 ################################################################################
 
-
 @pytest.fixture
```
Comment (Cadair) [inline ndcube/conftest.py:578]:
```suggestion

@pytest.fixture
```

File: ndcube/ndcube.py [hunk 2]
```diff
@@ -399,6 +399,42 @@ def __init__(self, data, wcs=None, uncertainty=None, mask=None, meta=None,
...
+        """
+        return super().data
+
+    @data.setter
+    def data(self, value):
+        # In an array-agnostic way check the shape is the same
+        if not hasattr(value, "shape") or value.shape != self.data.shape:
+            raise TypeError(f"Can only set data with an array-like object of the same shape ({self.data.shape})")
+
+        # Other masked arrays are hard to detect reliably
```
Comment (DanRyanIrish) [inline ndcube/ndcube.py:423]: So what happens if `value` a different type of masked array?  Does `.data` become that type of masked array?

File: ndcube/tests/test_ndcube.py [hunk 3]
```diff
@@ -1238,3 +1238,64 @@ def test_ndcube_quantity(ndcube_2d_ln_lt_units):
...
+
+
+def test_quantity_data_setter(ndcube_2d_ln_lt_units):
+    cube = ndcube_2d_ln_lt_units
+    assert cube.unit
+
+    new_data = np.zeros_like(cube.data) * cube.unit
+    cube.data = new_data
+
+    assert isinstance(cube.data, np.ndarray)
```
Comment (DanRyanIrish) [inline ndcube/tests/test_ndcube.py:1276]:
```suggestion
    assert isinstance(cube.data, np.ndarray)
    np.testing.assert_allclose(cube.data, new_data.value)
```

File: ndcube/ndcube.py (hunk 2 above)
Comment (Cadair) [reply ndcube/ndcube.py:423]: yes

Comment (Cadair) [reply ndcube/ndcube.py:423]: Probably the same thing.

Comment (Cadair) [reply ndcube/ndcube.py:423]: yes, the same thing:
```
In [10]: NDCube(data=dask.array.ma.zeros_like(np.array((10,10))), wcs=WCS()).data
Out[10]: dask.array<<numpy.ma.core._convert2ma object at 0x7f8144495c5, shape=(2,), dtype=int64, chunksize=(2,), chunktype=numpy.MaskedArray>
```

this is just unsupported anywhere.

File: ndcube/ndcube.py [hunk 4]
```diff
@@ -399,6 +399,42 @@ def __init__(self, data, wcs=None, uncertainty=None, mask=None, meta=None,
...
+        if isinstance(value, np.ma.MaskedArray):
```
[the remark this answers is not in the document]
Comment (DanRyanIrish) [reply ndcube/ndcube.py:424]: OK then.

**Review by DanRyanIrish — APPROVED** [review]

**Pull request merged**